### ISSUE TITLE
Run plan with two instances based on two different builds

### ISSUE THREAD

**Issue by mxinden** [testground/testground#1305]
:wave: libp2p steward here.

## Description

My larger goal is to test that go-libp2p and rust-libp2p are compatible.

As a first step I would like to:

- Run two instances, one based on go-libp2p, one based on rust-libp2p.
- Connecting to each other.
- Exchanging libp2p ping protocol messages.

I have a first proof-of-concept here: https://github.com/libp2p/test-plans/pull/20

Here I define two test plans, one Golang based, one Rust based:

- https://github.com/libp2p/test-plans/pull/20/files#diff-e604ac0c6c3c5649b5554487d4c69f1344a947bfb512f458af7c476d36e8378f
- https://github.com/libp2p/test-plans/pull/20/files#diff-13de5a9ebadea1c629f4768d0c538dcfdf3bd36b323e902f560add44d79359a8

I then define a *compositional* test plan which uses the outputs of the two standard plans above and overrides the `[groups.run] artifact` with the corresponding Docker images

https://github.com/libp2p/test-plans/pull/20/files#diff-2266c0d0d968091b0c145b03fb719583bc664631ccda4e844e07b1852e15cfe1

This is a very manual process. I first have to build both standard test plans, scan the output for the Docker image names, update the *compositional* test plan and only then execute it.

Can the same be achieved in an automatic fashion, excluding the manual steps? If not, is supporting this use-case on the roadmap?

Feedback on https://github.com/libp2p/test-plans/pull/20 directly is also much appreciated.

//CC @nonsense

**Comment by laurentsenta** [comment]
@mxinden It looks like there is no testground-native feature for composition right now, but we can automate your approach and reuse a few patterns found across the testground repo.

I built on top of your branch and poc'd automating composition files creation & dependency overrides:
https://github.com/libp2p/test-plans/pull/22/files#diff-4d2a8eefdf2a9783512a35da4dc7676a66404b6f3826a8af9aad038722da6823

I'm still working on it, but once I get the test to pass on a dev machine and the CI I'll probably need your input regarding the test content itself.

**Comment by mxinden** [comment]
@laurentsenta instead of a shell script, should this functionality not be a build-in feature of testground? I expect this feature to be useful beyond the libp2p compatibility use-case.

**Comment by laurentsenta** [comment]
@mxinden Definitely, I started looking at the builder implementation, it looks pretty doable, and I'm sure @nonsense might have some input on how to implement this.

I assume that we would get a lot of value from the feature _"A testground user can write, build, and run interop tests on their local machine and CI"_. So the sooner we get this feature, the sooner we can get its benefits and learn what other features we might need from the composition file.

My plan right now:

- Step 1: Automate your approach; it's clever; we can bypass the builder limitation, test the rest of the flow, and come back to the build step later,
- Step 2: Make sure the whole flow works (local dev and CI), get a few tests working this way,
- Step 3: Eventually, have this used by a few "early adopters" and identify other missing pieces,
- Step 4: Define a new composition file spec and implement it as a native feature,

Thanks for asking; that was not clear from my first comment; happy to get your feedback on this,

**Comment by mxinden** [comment]
Sounds good to me @laurentsenta. I will take a look at https://github.com/libp2p/test-plans/pull/22. Appreciate the help!

**Comment by laurentsenta** [comment]
Current PR with cross-version testing in libp2p: https://github.com/libp2p/test-plans/pull/23
(no interop with Rust yet, sorry @mxinden, I'm going back on this now)

**Comment by laurentsenta** [comment]
@mxinden now that #1337 is complete, I created https://github.com/testground/testground/issues/1357 to let you follow along the support for this feature.

**Comment by mxinden** [comment]
With https://github.com/testground/testground/issues/1357 merged, I think we can close here, right @laurentsenta?
